system-addon/content-src/components/TopSites/TopSite.jsx

```diff
@@ -175,24 +236,161 @@ export class TopSite extends React.PureComponent {
+  constructor(props) {
+    super(props);
+    this.state = this.DEFAULT_STATE = {
+      draggedIndex: null,
+      draggedSite: null,
+      draggedTitle: null,
+      topSitesPreview: null
+    };
+    this.onDragEvent = this.onDragEvent.bind(this);
+  }
+  componentWillUpdate(nextProps) {
+    if (this.state.draggedSite) {
+      const prevTopSites = this.props.TopSites && this.props.TopSites.rows;
+      const newTopSites = nextProps.TopSites && nextProps.TopSites.rows;
+      if (prevTopSites && prevTopSites[this.state.draggedIndex] &&
+        prevTopSites[this.state.draggedIndex].url === this.state.draggedSite.url &&
+        (!newTopSites[this.state.draggedIndex] || newTopSites[this.state.draggedIndex].url !== this.state.draggedSite.url)) {
+        // We got the new order from the redux store via props. We can clear state now.
+        this.setState(this.DEFAULT_STATE);
+      }
+    }
+  }
+  userEvent(event, index) {
+    this.props.dispatch(ac.UserEvent({
+      event,
+      source: TOP_SITES_SOURCE,
+      action_position: index
+    }));
+  }
+  onDragEvent(event, index, link, title) {
+    switch (event.type) {
+      case "dragstart":
+        this.onDragStart(index, link, title);
+        break;
+      case "dragend":
+        this.onDragEnd();
+        break;
+      case "dragover":
+      case "dragenter":
+        this.onDragEnter(index);
+        break;
+      case "dragleave":
+        this.onDragLeave();
+        break;
+      case "drop":
+        this.onDrop(index);
+        break;
+      default:
+        break;
+    }
+  }
+  onDragStart(index, link, title) {
+    this.setState({
+      draggedIndex: index,
+      draggedSite: link,
+      draggedTitle: title
+    });
+    this.userEvent("DRAG", index);
+  }
+  onDragEnd() {
+    this.setState(this.DEFAULT_STATE);
+  }
+  onDragEnter(index) {
+    this.setState({topSitesPreview: this._makeTopSitesPreview(index)});
+  }
+  onDragLeave() {
+    this.setState({topSitesPreview: null});
+  }
+  onDrop(index) {
+    this.props.dispatch(ac.SendToMain({
+      type: at.TOP_SITES_INSERT,
+      data: {site: {url: this.state.draggedSite.url, label: this.state.draggedTitle}, index}
+    }));
+    this.userEvent("DROP", index);
+  }
+  _getTopSites() {
+    return this.props.TopSites.rows.slice(0, this.props.TopSitesCount);
+  }
+
+  /**
+   * Make a preview of the topsites that will be the result of dropping the currently
+   * dragged site at the specified index.
+   */
+  _makeTopSitesPreview(index) {
+    const preview = this._getTopSites();
+    this._fillOrLeaveHole(preview, this.state.draggedIndex);
+    this._insertSite(preview, Object.assign({}, this.state.draggedSite, {isPinned: true}), index);
+    return preview;
+  }
+
+  /**
+   * Fill in the slot at the specified index with a non pinned site further down the
+   * list, if any. Otherwise leave an empty slot.
+   */
+  _fillOrLeaveHole(sites, index) {
+    let slotIndex = index;
```

Comment:
Could just define the method to have `slotIndex` as its parameter? ;)

Reply:
this was because linter was complaining about reassigning to index. I think it looks funny this way though. already using index and i in the function.